packages/omgidl-serialization/src/MessageReader.ts

```diff
@@ -82,109 +82,214 @@ export class MessageReader<T = unknown> {
-              field.name
-            }" in ${complexDef.name ?? "unknown"}`,
-          );
-        }
+
+      msg[field.name] = this.readMemberFieldValue(
+        field,
+        reader,
+        { readMemberHeader, parentName: complexDef.name },
+        childOptions,
+      );
+    }
+
+    return msg;
+  }
+
+  private readUnionType(
+    unionDef: IDLUnionDefinition,
+    reader: CdrReader,
+    options: {
+      isTopLevel: boolean;
+      usesDelimiterHeader: boolean;
+      usesMemberHeader: boolean;
+    },
+  ): Record<string, unknown> {
+    const { usesMemberHeader } = options;
+    const { typeUsesMemberHeader } = getHeaderNeeds(unionDef);
+
+    const readMemberHeader = usesMemberHeader && typeUsesMemberHeader;
+
+    const childOptions = {
+      ...options,
+      isTopLevel: false,
+    };
+
+    // read switchtype value
+    const switchTypeDeser = deserializers.get(unionDef.switchType);
+    if (switchTypeDeser == undefined) {
+      throw new Error(`Unrecognized switch discriminator type ${unionDef.switchType}`);
+    }
+
+    // looks like unions print an emHeader for the switchType
+    if (readMemberHeader) {
+      const { objectSize: objectSizeBytes } = reader.emHeader();
+      const switchTypeLength = typeToByteLength(unionDef.switchType);
+      if (switchTypeLength != undefined && objectSizeBytes !== switchTypeLength) {
+        throw new Error(
+          `Expected switchType length of ${switchTypeLength} but got ${objectSizeBytes} for ${
+            unionDef.name ?? ""
+          }`,
+        );
       }
+    }
+    const discriminatorValue = switchTypeDeser(reader) as number | boolean;
 
-      if (field.isComplex === true) {
-        // Complex type
-        const nestedComplexDef = this.definitions.get(field.type);
-        if (nestedComplexDef == undefined) {
-          throw new Error(`Unrecognized complex type ${field.type}`);
-        }
+    // get case for switchtype value based on matching predicate
 
-        if (field.isArray === true) {
-          // For dynamic length arrays we need to read a uint32 prefix
-          const arrayLengths = field.arrayLengths ?? [reader.sequenceLength()];
+    const caseDefType = getCaseForDiscriminator(unionDef, discriminatorValue);
+    if (!caseDefType) {
+      throw new Error(
+        `No matching case found in ${
+          unionDef.name ?? ""
+        } for discriminator value ${discriminatorValue.toString()}`,
+      );
+    }
 
-          const complexDeserializer = () => {
-            return this.readComplexType(nestedComplexDef, reader, childOptions);
+    // read case value
+    const msg: Record<string, unknown> = {};
+    msg[caseDefType.name] = this.readMemberFieldValue(
+      caseDefType,
+      reader,
+      { readMemberHeader, parentName: unionDef.name },
+      childOptions,
+    );
+    return msg;
+  }
+
+  private readMemberFieldValue(
```

Comment:
note most of this is the same as it was previously in `readComplexType` with exception of `readAggregatedType` and it returning the value of the field instead of placing it directly into the message